sklearn/linear_model/tests/test_coordinate_descent.py

```diff
@@ -804,3 +814,61 @@ def test_enet_l1_ratio():
         est.fit(X, y[:, None])
         est_desired.fit(X, y[:, None])
     assert_array_almost_equal(est.coef_, est_desired.coef_, decimal=5)
+
+
+def test_enet_duality_gap():
+    n_samples = 100
+    n_features = 500
+    rng = np.random.RandomState(42)
+    kwargs = {}
+    if sp_version >= (0, 13):
+        kwargs['random_state'] = 42
+    X_csc = sparse.random(n_samples, n_features, density=0.3, **kwargs).tocsc()
+    X_csc = X_csc.astype(np.float32)
+    X = np.asfortranarray(X_csc.toarray())
+
+    def duality_gap(X, y, w, alpha, beta, positive=False):
+        """Slow numpy based formula for the duality gap of elasticnet"""
+        # TODO: implement positive=True case
+        R = y - np.dot(X, w)
+        primal = (0.5 * (R ** 2).sum()
+                  + alpha * np.abs(w).sum()
+                  + 0.5 * beta * (w ** 2).sum())
+        # realizable dual solution
+        theta = R / max(alpha, np.abs(np.dot(X.T, R)).max())
+        dual = 0.5 * ((y ** 2).sum()
+                      - alpha ** 2 * ((y / alpha - theta) ** 2).sum()
+                      - alpha ** 2 * beta * ((w ** 2).sum()))
```

Comment:
After investigation, there is a mismatch between the gap from cd_fast and gap = primal - dual.
Someone can confirm that the dual formula is correct ? @josephsalmon @agramfort ?